File: android/app/src/main/java/de/wolmanager/data/Repo.kt
package de.wolmanager.data

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.io.File
import java.util.UUID

/**
 * JSON-Repository (devices/schedules/logs/settings) unter [baseDir].
 * devices.json bleibt ein JSON-ARRAY (Desktop-Kompatibilität). Passwörter werden
 * vor dem Schreiben geleert und an [secure] delegiert.
 */
class Repo(
    private val baseDir: File,
    private val secure: SecureStore?,
) {
    private val json = Json {
        prettyPrint = true
        ignoreUnknownKeys = true
        encodeDefaults = true
    }

    data class Snapshot(
        val devices: List<Device> = emptyList(),
        val schedules: List<ScheduleDef> = emptyList(),
        val logs: List<LogEntry> = emptyList(),
        val settings: AppSettings = AppSettings(),
    )

    private val _snapshot = MutableStateFlow(load())
    val snapshot: StateFlow<Snapshot> = _snapshot.asStateFlow()

    private val devicesFile get() = File(baseDir, "devices.json")
    private val schedulesFile get() = File(baseDir, "schedules.json")
    private val logsFile get() = File(baseDir, "logs.json")
    private val settingsFile get() = File(baseDir, "settings.json")

    // ── Init ────────────────────────────────────────────────────────────────

    private fun load(): Snapshot {
        baseDir.mkdirs()
        val devices = (read(devicesFile) { json.decodeFromString<List<Device>>(it) } ?: emptyList())
            .map { migratePasswords(it) }
        val schedules = read(schedulesFile) { json.decodeFromString<List<ScheduleDef>>(it) } ?: emptyList()
        val logs = read(logsFile) { json.decodeFromString<List<LogEntry>>(it) } ?: emptyList()
        val settings = read(settingsFile) { json.decodeFromString<AppSettings>(it) } ?: AppSettings()
        return Snapshot(devices, schedules, logs, settings)
    }

    /** Einmalige Migration: Klartext-Passwort aus devices.json in den SecureStore ziehen. */
    private fun migratePasswords(d: Device): Device {
        if (d.password.isNotEmpty() && secure != null) {
            val existing = secure.getPassword(d.id)
            if (existing.isEmpty()) secure.setPassword(d.id, d.password)
            return d.copy(password = "")
        }
        return d
    }

    private inline fun <T> read(file: File, block: (String) -> T): T? = try {
        if (file.exists()) block(file.readText()) else null
    } catch (_: Exception) {
        null
    }

    private inline fun <reified T : Any> write(file: File, value: T, serializer: (T) -> String) {
        try {
            baseDir.mkdirs()
            val tmp = File(file.parentFile, file.name + ".tmp")
            tmp.writeText(serializer(value))
            // renameTo überschreibt auf Windows nicht → Ziel zuerst entfernen.
            if (!tmp.renameTo(file)) {
                runCatching { file.delete() }
                tmp.renameTo(file)
            }
        } catch (_: Exception) {
        }
    }

    // ── Geräte ──────────────────────────────────────────────────────────────

    private fun persistDevices(devices: List<Device>) {
        write(devicesFile, devices) { json.encodeToString(it) }
    }

    fun saveDevice(device: Device) {
        val pw = device.password
        val id = device.id.ifEmpty { UUID.randomUUID().toString() }
        val stored = device.copy(id = id, password = "")
        secure?.setPassword(id, pw)
        val cur = _snapshot.value.devices.toMutableList()
        val idx = cur.indexOfFirst { it.id == id }
        if (idx >= 0) cur[idx] = stored else cur.add(stored)
        persistDevices(cur)
        _snapshot.value = _snapshot.value.copy(devices = cur)
    }

    fun deleteDevice(id: String) {
        val cur = _snapshot.value.devices.filter { it.id != id }
        secure?.removePassword(id)
        persistDevices(cur)
        // Zugehörige Zeitpläne mit entfernen
        val sched = _snapshot.value.schedules.filter { it.deviceId != id }
        write(schedulesFile, sched) { json.encodeToString(it) }
        _snapshot.value = _snapshot.value.copy(devices = cur, schedules = sched)
    }

    fun getPassword(id: String): String = secure?.getPassword(id) ?: ""

    fun deviceName(id: String): String? =
        _snapshot.value.devices.firstOrNull { it.id == id }?.name

    // ── Zeitpläne ───────────────────────────────────────────────────────────

    fun saveSchedule(schedule: ScheduleDef) {
        val id = schedule.id.ifEmpty { UUID.randomUUID().toString() }
        val stored = schedule.copy(id = id)
        val cur = _snapshot.value.schedules.toMutableList()
        val idx = cur.indexOfFirst { it.id == id }
        if (idx >= 0) cur[idx] = stored else cur.add(stored)
        write(schedulesFile, cur) { json.encodeToString(it) }
        _snapshot.value = _snapshot.value.copy(schedules = cur)
    }

    fun deleteSchedule(id: String) {
        val cur = _snapshot.value.schedules.filter { it.id != id }
        write(schedulesFile, cur) { json.encodeToString(it) }
        _snapshot.value = _snapshot.value.copy(schedules = cur)
    }

    fun updateScheduleLastRun(id: String, ts: Long) {
        val cur = _snapshot.value.schedules.map {
            if (it.id == id) it.copy(lastRun = ts) else it
        }
        write(schedulesFile, cur) { json.encodeToString(it) }
        _snapshot.value = _snapshot.value.copy(schedules = cur)
    }

    // ── Protokoll ───────────────────────────────────────────────────────────

    fun log(device: String, level: String, msg: String) {
        val entry = LogEntry(System.currentTimeMillis(), device, level, msg)
        val max = _snapshot.value.settings.maxLogs.coerceAtLeast(10)
        val cur = (listOf(entry) + _snapshot.value.logs).take(max)
        write(logsFile, cur) { json.encodeToString(it) }
        _snapshot.value = _snapshot.value.copy(logs = cur)
    }

    fun clearLogs() {
        write(logsFile, emptyList<LogEntry>()) { json.encodeToString(it) }
        _snapshot.value = _snapshot.value.copy(logs = emptyList())
    }

    // ── Einstellungen ───────────────────────────────────────────────────────

    fun saveSettings(settings: AppSettings) {
        write(settingsFile, settings) { json.encodeToString(it) }
        _snapshot.value = _snapshot.value.copy(settings = settings)
    }

    fun resetSettings() = saveSettings(AppSettings())
}

File: ios/WolManager/Model/Repo.swift
import Foundation

/*
 * JSON-Repository (devices/schedules/logs/settings) unter [baseDir].
 * devices.json bleibt ein JSON-ARRAY (Desktop-Kompatibilität). Passwörter werden
 * vor dem Schreiben geleert und an SecureStore (Keychain) delegiert.
 * Analog zu Repo.kt — atomares Schreiben über .tmp + Rename.
 */
final class Repo {

    struct Snapshot {
        var devices: [Device] = []
        var schedules: [ScheduleDef] = []
        var logs: [LogEntry] = []
        var settings = AppSettings()
    }

    /// Wird nach jeder Änderung mit dem neuen Snapshot aufgerufen (Main-Queue).
    var onChange: ((Snapshot) -> Void)?

    private let baseDir: URL
    private let secure: SecureStore
    private let lock = NSLock()
    private var state: Snapshot

    private let enc: JSONEncoder = {
        let e = JSONEncoder()
        e.outputFormatting = [.prettyPrinted, .sortedKeys]
        return e
    }()
    private let dec = JSONDecoder()

    init(baseDir: URL, secure: SecureStore) {
        self.baseDir = baseDir
        self.secure = secure
        self.state = Snapshot()
        reload()
    }

    private var devicesFile: URL { baseDir.appendingPathComponent("devices.json") }
    private var schedulesFile: URL { baseDir.appendingPathComponent("schedules.json") }
    private var logsFile: URL { baseDir.appendingPathComponent("logs.json") }
    private var settingsFile: URL { baseDir.appendingPathComponent("settings.json") }

    // ── Init ────────────────────────────────────────────────────────────────

    func reload() {
        try? FileManager.default.createDirectory(at: baseDir, withIntermediateDirectories: true)
        var snap = Snapshot()
        snap.devices = (readList(devicesFile, Device.self) ?? []).map { migratePasswords($0) }
        snap.schedules = readList(schedulesFile, ScheduleDef.self) ?? []
        snap.logs = readList(logsFile, LogEntry.self) ?? []
        snap.settings = readOne(settingsFile, AppSettings.self) ?? AppSettings()
        lock.lock()
        state = snap
        lock.unlock()
    }

    /// Einmalige Migration: Klartext-Passwort aus devices.json in den Keychain ziehen.
    private func migratePasswords(_ d: Device) -> Device {
        if !d.password.isEmpty {
            if secure.getPassword(deviceId: d.id).isEmpty {
                secure.setPassword(deviceId: d.id, password: d.password)
            }
            var copy = d
            copy.password = ""
            return copy
        }
        return d
    }

    private func readList<T: Decodable>(_ file: URL, _ type: T.Type) -> [T]? {
        guard let text = try? String(contentsOf: file, encoding: .utf8) else { return nil }
        return try? dec.decode([T].self, from: Data(text.utf8))
    }

    private func readOne<T: Decodable>(_ file: URL, _ type: T.Type) -> T? {
        guard let text = try? String(contentsOf: file, encoding: .utf8) else { return nil }
        return try? dec.decode(T.self, from: Data(text.utf8))
    }

    private func write<T: Encodable>(_ file: URL, _ value: T) {
        do {
            try FileManager.default.createDirectory(at: baseDir, withIntermediateDirectories: true)
            let data = try enc.encode(value)
            let tmp = baseDir.appendingPathComponent(file.lastPathComponent + ".tmp")
            try data.write(to: tmp, options: .atomic)
            // replaceItem überschreibt atomar; Ziel zuerst entfernen, falls nicht vorhanden.
            if FileManager.default.fileExists(atPath: file.path) {
                _ = try? FileManager.default.replaceItemAt(file, withItemAt: tmp)
            } else {
                try FileManager.default.moveItem(at: tmp, to: file)
            }
        } catch {
            NSLog("Repo.write failed for \(file.lastPathComponent): \(error)")
        }
    }

    // ── Zugriff ─────────────────────────────────────────────────────────────

    var snapshot: Snapshot {
        lock.lock(); defer { lock.unlock() }
        return state
    }

    private func notify() {
        let snap = snapshot
        DispatchQueue.main.async { [onChange] in onChange?(snap) }
    }

    // ── Geräte ──────────────────────────────────────────────────────────────

    @discardableResult
    func saveDevice(_ device: Device) -> Device {
        let pw = device.password
        let id = device.id.isEmpty ? UUID().uuidString : device.id
        var stored = device
        stored.id = id
        stored.password = ""
        secure.setPassword(deviceId: id, password: pw)
        lock.lock()
        var cur = state.devices
        if let idx = cur.firstIndex(where: { $0.id == id }) { cur[idx] = stored } else { cur.append(stored) }
        state.devices = cur
        lock.unlock()
        write(devicesFile, cur)
        notify()
        return stored
    }

    func deleteDevice(id: String) {
        secure.removePassword(deviceId: id)
        lock.lock()
        let cur = state.devices.filter { $0.id != id }
        let sched = state.schedules.filter { $0.deviceId != id }
        state.devices = cur
        state.schedules = sched
        lock.unlock()
        write(devicesFile, cur)
        write(schedulesFile, sched)
        notify()
    }

    func getPassword(id: String) -> String { secure.getPassword(deviceId: id) }

    func deviceName(id: String) -> String? {
        snapshot.devices.first(where: { $0.id == id })?.name
    }

    func device(_ id: String) -> Device? {
        snapshot.devices.first(where: { $0.id == id })
    }

    // ── Zeitpläne ───────────────────────────────────────────────────────────

    @discardableResult
    func saveSchedule(_ schedule: ScheduleDef) -> ScheduleDef {
        let id = schedule.id.isEmpty ? UUID().uuidString : schedule.id
        var stored = schedule
        stored.id = id
        lock.lock()
        var cur = state.schedules
        if let idx = cur.firstIndex(where: { $0.id == id }) { cur[idx] = stored } else { cur.append(stored) }
        state.schedules = cur
        lock.unlock()
        write(schedulesFile, cur)
        notify()
        return stored
    }

    func deleteSchedule(id: String) {
        lock.lock()
        let cur = state.schedules.filter { $0.id != id }
        state.schedules = cur
        lock.unlock()
        write(schedulesFile, cur)
        notify()
    }

    func updateScheduleLastRun(id: String, ts: Int64) {
        lock.lock()
        let cur = state.schedules.map { $0.id == id ? ScheduleDef(id: $0.id, deviceId: $0.deviceId,
            hour: $0.hour, minute: $0.minute, days: $0.days, enabled: $0.enabled,
            action: $0.action, lastRun: ts) : $0 }
        state.schedules = cur
        lock.unlock()
        write(schedulesFile, cur)
        notify()
    }

    // ── Protokoll ───────────────────────────────────────────────────────────

    func log(device: String, level: String, msg: String) {
        let entry = LogEntry(ts: Int64(Date().timeIntervalSince1970 * 1000), device: device, level: level, msg: msg)
        lock.lock()
        let max = max(state.settings.maxLogs, 10)
        let cur = ([entry] + state.logs).prefix(max)
        state.logs = Array(cur)
        lock.unlock()
        write(logsFile, state.logs)
        notify()
    }

    func clearLogs() {
        lock.lock()
        state.logs = []
        lock.unlock()
        write(logsFile, [LogEntry]())
        notify()
    }

    // ── Einstellungen ───────────────────────────────────────────────────────

    func saveSettings(_ settings: AppSettings) {
        lock.lock()
        state.settings = settings
        lock.unlock()
        write(settingsFile, settings)
        notify()
    }

    func resetSettings() { saveSettings(AppSettings()) }
}
